runtime error: slice bounds out of range in packets.go

### Issue description
When using many concurrent SQL queries (same pool), I get a slice out of bounds error.
The corresponding logic in [jmoiron/sqlx](https://github.com/jmoiron/sqlx) seems fine. It's just a call to `(*Rows).Next()` which is causing this.

Similar to #36 perhaps?

### Error log
```
panic: runtime error: slice bounds out of range

goroutine 4814 [running]:
mypackage/vendor/github.com/go-sql-driver/mysql.(*binaryRows).readRow(0xc4205a2160, 0xc422be3140, 0xc, 0xc, 0xc, 0xc422be3080)
	/home/gm42/projects/mypackage/src/mypackage/vendor/github.com/go-sql-driver/mysql/packets.go:1168 +0x92d
mypackage/vendor/github.com/go-sql-driver/mysql.(*binaryRows).Next(0xc4205a2160, 0xc422be3140, 0xc, 0xc, 0xc, 0xc)
	/home/gm42/projects/mypackage/src/mypackage/vendor/github.com/go-sql-driver/mysql/rows.go:85 +0x5e
database/sql.(*Rows).nextLocked(0xc422b6da80, 0xc4206ac4b0)
	/usr/local/go/src/database/sql/sql.go:2149 +0x6c
database/sql.(*Rows).Next.func1()
	/usr/local/go/src/database/sql/sql.go:2134 +0x3c
database/sql.withLock(0xeb1640, 0xc422b6dab0, 0xc4206ac4f0)
	/usr/local/go/src/database/sql/sql.go:2545 +0x65
database/sql.(*Rows).Next(0xc422b6da80, 0xb82ce0)
	/usr/local/go/src/database/sql/sql.go:2135 +0x83
mypackage/vendor/github.com/jmoiron/sqlx.(*Row).Scan(0xc4223b0870, 0xc422be2fc0, 0xc, 0xc, 0x0, 0x0)
```

### Configuration
*Driver version (or git SHA):* 1421caf44f6464fd2ee8de694c7508ee13f92964

*Go version:* go version go1.8.1 linux/amd64

*Server version:* MySQL 5.6.27

*Server OS:* Ubuntu 16

I've been able to reproduce this on my laptop with @gm42 submitted test case.

mysql version: mysqld  Ver 5.7.18 for osx10.12 on x86_64 (Homebrew)

go version: 1.8.1
go-sql-driver: latest master
sqlx: latest master

result: it seems to fail randomly in a different iteration

```
2017/05/10 11:53:14 starting iterations
2017/05/10 11:53:14 TestcaseIteration(40): NOT A BUG
2017/05/10 11:53:14 TestcaseIteration(41): 67108864
panic: BUG

```
```
2017/05/10 11:53:56 starting iterations
2017/05/10 11:53:56 TestcaseIteration(45): NOT A BUG
2017/05/10 11:53:56 TestcaseIteration(46): NOT A BUG
2017/05/10 11:53:56 TestcaseIteration(48): NOT A BUG
2017/05/10 11:53:56 TestcaseIteration(47): NOT A BUG
2017/05/10 11:53:56 TestcaseIteration(43): NOT A BUG
2017/05/10 11:53:56 TestcaseIteration(42): 67108864
panic: BUG

```

--- a/rows.go
+++ b/rows.go
@@ -33,12 +33,6 @@ type mysqlRows struct {
 
 type binaryRows struct {
 	mysqlRows
-	// stmtCols is a pointer to the statement's cached columns for different
-	// result sets.
-	stmtCols *[][]mysqlField
-	// i is a number of the current result set. It is used to fetch proper
-	// columns from stmtCols.
-	i int
 }
 
 type textRows struct {
@@ -132,25 +126,14 @@ func (rows *mysqlRows) nextNotEmptyResultSet() (int, error) {
 	}
 }
 
-func (rows *binaryRows) NextResultSet() (err error) {
+func (rows *binaryRows) NextResultSet() error {
 	resLen, err := rows.nextNotEmptyResultSet()
 	if err != nil {
 		return err
 	}
 
-	// get columns, if not cached, read them and cache them.
-	if rows.i >= len(*rows.stmtCols) {
-		rows.rs.columns, err = rows.mc.readColumns(resLen)
-		*rows.stmtCols = append(*rows.stmtCols, rows.rs.columns)
-	} else {
-		rows.rs.columns = (*rows.stmtCols)[rows.i]
-		if err := rows.mc.readUntilEOF(); err != nil {
-			return err
-		}
-	}
-
-	rows.i++
-	return nil
+	rows.rs.columns, err = rows.mc.readColumns(resLen)
+	return err
 }
 
 func (rows *binaryRows) Next(dest []driver.Value) error {

--- a/statement.go
+++ b/statement.go
@@ -20,7 +20,6 @@ type mysqlStmt struct {
 	mc         *mysqlConn
 	id         uint32
 	paramCount int
-	columns    [][]mysqlField // cached from the first query
 }
 
 func (stmt *mysqlStmt) Close() error {
@@ -109,20 +108,10 @@ func (stmt *mysqlStmt) Query(args []driver.Value) (driver.Rows, error) {
 	}
 
 	rows := new(binaryRows)
-	rows.stmtCols = &stmt.columns
 
 	if resLen > 0 {
 		rows.mc = mc
-		rows.i++
-		// Columns
-		// If not cached, read them and cache them
-		if len(stmt.columns) == 0 {
-			rows.rs.columns, err = mc.readColumns(resLen)
-			stmt.columns = append(stmt.columns, rows.rs.columns)
-		} else {
-			rows.rs.columns = stmt.columns[0]
-			err = mc.readUntilEOF()
-		}
+		rows.rs.columns, err = mc.readColumns(resLen)
 	} else {
 		rows.rs.done = true
 

